File: src/main/java/linq/collections/enumerable/ArrayEnumerable.java
package linq.collections.enumerable;

import linq.Enumerable;
import linq.Enumerator;
import linq.collections.enumerator.ArrayEnumerator;

public class ArrayEnumerable<T> extends Enumerable<T> {
    private final T[] source;

    public ArrayEnumerable(T[] source) {
        this.source = source;
    }

    public T get(int index) {
        return source[index];
    }

    @Override
    public Enumerator<T> enumerator() {
        return new ArrayEnumerator<>(source);
    }

    @Override
    public int count() {
        return source.length;
    }
}

File: src/main/kotlin/linq/collections/enumerable/ArrayEnumerable.kt
package linq.collections.enumerable

import linq.Enumerable
import linq.Enumerator
import linq.collections.enumerator.ArrayEnumerator

class ArrayEnumerable<T>(private val source: Array<out T>) : Enumerable<T>() {
    val size: Int = source.size

    operator fun get(index: Int): T = source[index]

    override fun enumerator(): Enumerator<T> = ArrayEnumerator(source)
}